napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,138 @@
+
+logger = logging.getLogger("napari.components._layer_slicer")
+
+
+class _LayerSlicer:
+    """
+    High level class to control the creation of a slice (via a slice request),
+    submit it (synchronously or asynchronously) to a thread pool, and emit the
+    results when complete.
+
+    Events
+    ------
+    ready
+        emitted after slicing is done with a dict value that maps from layer to slice response
+    """
+
+    def __init__(self):
+        """
+        Attributes
+        ----------
+        _executor : concurrent.futures.ThreadPoolExecutor
+            manager for the slicing threading
+        _layers_to_task : dict
+            task storage for cancellation logic
+        _lock_layers_to_task : threading.RLock
+            lock to guard against changes to `_layers_to_task` when finding,
+            adding, or removing tasks.
+        """
+        self.events = EmitterGroup(source=self, ready=Event)
+        self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._layers_to_task: Dict[Tuple[Layer], Future] = {}
+        self._lock_layers_to_task = RLock()
+
+    def slice_layers_async(
+        self, layers: Iterable[Layer], dims: Dims
+    ) -> Future[dict]:
+        """This should only be called from the main thread.
+
+        Creates a new task and adds it to the _layers_to_task dict. Cancels
+        all tasks currently pending for that layer tuple.
+
+        Submitting multiple layers at one generates multiple requests (stored in a dict),
+        but only ONE task.
+
+        A slice submitted with multiple layers will only cancel if the same tuple of layers
+        is resubmitted. If a second slice is requested with only one layer from the first
+        slice, it will not be cancelled.
+        """
+        # Cancel any tasks that are slicing a subset of the layers
+        # being sliced now. This allows us to slice arbitrary sets of
+        # layers with some sensible and not too complex cancellation
+        # policy.
+        if existing_task := self._find_existing_task(layers):
+            logger.debug('Cancelling task for %s', layers)
+            existing_task.cancel()
+
+        # Not all layer types will initially be asynchronously sliceable.
+        # The following logic gives us a way to handle those in the short
+        # term as we develop, and also in the long term if there are cases
+        # when we want to perform sync slicing anyway.
+        requests = {}
+        for layer in layers:
+            if layer._is_async():
+                requests[layer] = layer._make_slice_request(dims)
+            else:
+                layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+        # create task for slicing of each request/layer
+        task = self._executor.submit(self._slice_layers, requests)
+
+        # store task for cancellation logic
+        self._layers_to_task[tuple(requests.keys())] = task
+
+        task.add_done_callback(self._on_slice_done)
```

Comment:
Yes, in most interesting cases this won't be run on the main thread. Though in some cases it will be (e.g. cancellation). That's indicated in the docstring of `_on_slice_done`. 

But I think that's fine, as anyone using this private class directly should be expected to reason about multiple threads.

We should clarify that `_LayerSlicer.events.ready` can (and will often) be emitted on a non-main thread. But again anyone using this private class directly (which should only be napari developers) should do the work to ensure that their particular callback is executed on the main thread (e.g. by decorating their callback with ensure_main_thread` as in the prototype.

@kcpevey : can you add something to the class docstring to clarify that expecation?